README.md
Webclient for 10x10 draughts
----------------------------
Play against your favorite draughts engine like MobyDam. <br/>
Move pieces on the board by point and click. <br/>
This browser client and a draughts engine can be directly connected if the engine uses the Websocket protocol.  <br/>

![connect](include/images/connect.png)

Often draughts engines use the Tcp Socket protocol. But browser clients do not understand this protocol. <br/>
To solve this problem an intermediate server has to put between client and engine.  <br/>
This server acts as a bridge for the messages between the client and the engine. <br/>

![bridge](include/images/bridge.png)

One great advantage of using a browser client is that you are independent of the operating system like
Linux, Windows and Mac. <br/>
The price you have to pay is the use of a bridge server. <br/>
Until draughts engines are designed to communicate directly with a browser client. <br/>
But it does not cost you anything, because the bridge server application is free to download. <br/>

Links
-----
Use these links to setup an environment where you can play against the draughts engine MobyDam.
- [Github MobyDam Draughts engine](https://github.com/rhalbersma/mobydam)
- [Github Web2Tcp Bridge Server](https://github.com/akalverboer/web2tcp_bridge)
- [Website WsDam webclient](http://svg_experimenten.deds.nl/wsdam/wsdam.html)

<hr/>


CHANGE_LOG.md
Change log
----------
2018-04-25: Initial release <br/>

2018-04-30:
- Stopwatch for measuring thinking time between moves. <br/>
- Setup mode to put pieces with mouse clicks. <br/>
- Some general improvements. <br/>


include/js/wsdam_main.js
/*
||====================================================================================  
||  WsDam: browser client for 10x10 draughts
||====================================================================================
||  Main module.
||
|| (c) Copyright 2018 Arthur Kalverboer 
||
||====================================================================================
*/

function isLowerCase(str) {
   return str == str.toLowerCase() && str != str.toUpperCase();
}

function isUpperCase(str) {
   return str == str.toUpperCase() && str != str.toLowerCase();
}

function swapCase(char) {
   // Swap to upper or lower char.
   res = isLowerCase(char) ? char.toUpperCase() : char.toLowerCase()
   return res;
}  // swapCase()

String.prototype.lpad = function(padString, length) {
   // pads left
   var str = this.slice(0,length);
   while (str.length < length)
      str = padString + str;
   return str;
}  // lpad
 
String.prototype.rpad = function(padString, length) {
   // pads right
   var str = this.slice(0,length);
   while (str.length < length)
      str = str + padString;
   return str;
}  // rpad()

function numberToRowCol(m) {
   // Convert square number to row/column values depending on board dimension.
   // Numbering and row/col of board from top-left to bottom-right.
   if (C.BOARD_DIM == 8) {  // 8x8 board
      var col = ( (m-1) % 4 ) * 2  + ( ( Math.floor((m-1) / 4) ) + 1) % 2;
      var row = Math.floor( (m-1) / 4 );
   }
   else if (C.BOARD_DIM == 10) {     // 10x10 board
      var col = ( (m-1) % 5 ) * 2  + ( ( Math.floor((m-1) / 5) ) + 1) % 2;
      var row = Math.floor( (m-1) / 5 );
   }
   else if (C.BOARD_DIM == 12) {     // 12x12 board
      var col = ( (m-1) % 6 ) * 2  + ( ( Math.floor((m-1) / 6) ) + 1) % 2;
      var row = Math.floor( (m-1) / 6 );
   }
   else {
      var col = row = 0;  // unknown dim
   }
   return {row: row, col: col};
}  // numberToRowCol()

function rowcolToNumber(row, col) {
   // Converts row/col values to the square number depending on board dimension.
   // Numbering and row/col of board from top-left to bottom-right.
   if (C.BOARD_DIM == 8) {           // 8x8 board
      var m = (row * 4) + Math.ceil( (col + 1) / 2 );
   }
   else if (C.BOARD_DIM == 10) {     // 10x10 board
      var m = (row * 5) + Math.ceil( (col + 1) / 2 );
   }
   else if (C.BOARD_DIM == 12) {     // 12x12 board
      var m = (row * 6) + Math.ceil( (col + 1) / 2 );
   }
   else {
      var m = 0; // unknown dim
   }
   return m;
} // rowcolToNumber()


function arraysEqual(a1,a2, doSort) {
   // If sorted: equality as sets. Not sorted: equality as arrays.
   // Default doSort: true.
   // WARNING: arrays must NOT contain {objects} or behavior may be undefined.
   if (doSort) {
      a1_sorted = a1.slice(0).sort();  // clone and sort
      a2_sorted = a2.slice(0).sort();  // clone and sort
      return JSON.stringify(a1_sorted) == JSON.stringify(a2_sorted);
   }
   return JSON.stringify(a1) == JSON.stringify(a2);
}  // arraysEqual

function isSuperset(arr1, arr2) {
   // Returns true if the elements of array arr1 is a superset of arr2.
   // Example 1: [1, 'a', 2, 'b', 3]  is superset of [1, 2, 3]
   // Example 2: ['a', 2, 'b', 3] is NOT superset of [1, 2, 3]
   var bool = arr2.every(function (val) { return arr1.indexOf(val) >= 0; });
   return bool;
}  // isSuperset

window.StopWatch = function() {
   this.startMilliSec = 0;
   this.elapsedMilliSec = 0;
   this.start = function() {
      this.startMilliSec = new Date().getTime();
   }
   this.stop = function() {
      this.elapsedMilliSec = new Date().getTime() - this.startMilliSec;
   }
   return this;
}  // StopWatch()

window.CanvasHandler = function() {
   this.canvas = document.getElementById('myCanvas'); 
   this.canvas.width = C.CANVAS_SIZE.w;
   this.canvas.height = C.CANVAS_SIZE.h;

   this.ctx = this.canvas.getContext('2d'); 
   var this2 = this;  // Increase scope of this

   // === METHODS ===
   this.initEventListener = function() {
      // Init by pages who needs mouseclick-events on a square of the canvas board.
      function getX(evt) {
         return evt.offsetX ? evt.offsetX : evt.layerX;
      }  // getX

      function getY(evt) {
         return evt.offsetY ? evt.offsetY : evt.layerY;
      }  // getY

      function onMousedown(evt) {
         // Executes function 'onSquareClick' with argument square position.
         // Square position is (row,col) object.
         // Does nothing if no square selected.
         var button = {left: 0, middle: 1, right: 2};
         if (evt.button === button.right) { return null; }    // skip right mouse click
         if (evt.button === button.middle) { return null; }   // skip middle mouse click

         var mousePos = {x: getX(evt), y: getY(evt)};
         if (mousePos == null) { return null; }   // No data

         var bx = C.BOARD_RECT.x;
         var by = C.BOARD_RECT.y;
         var origin = { x: bx, y: by };         // origin of board

         var col = Math.floor( (mousePos.x - origin.x) / C.SQUARESIZE );  // 0..9
         var row = Math.floor( (mousePos.y - origin.y) / C.SQUARESIZE );  // 0..9
         if ((col < 0) || (col > (C.BOARD_DIM - 1) )) return null;
         if ((row < 0) || (row > (C.BOARD_DIM - 1) )) return null;

         if (isEven(row + col) == true) return null;     // remove for chess
         var pos = {row: row, col: col};   // coordinates of clicked square

         // Process 'mousedown' on a square; give mouse-position and which keys are pressed.
         if (typeof game.onSquareClick === 'function') { game.onSquareClick(pos, keysPressed); }
         return 0;
      }  // onMousedown

      var isEven = function(someNumber) { return (someNumber%2 == 0) ? true : false; };

      var keysPressed = {};   // remember which keys are pressed
      document.addEventListener('keydown', function(e) { keysPressed[e.keyCode] = true; }, false);
      document.addEventListener('keyup', function(e) { keysPressed[e.keyCode] = false;  }, false);
      this.canvas.addEventListener('mousedown', onMousedown, false);
      return 0;
   }  // initEventListener

   return this;
}  // CanvasHandler()

window.Game = function() {
   // Game logic. "Two-Color" Logic (moves and position).

   this.color = C.WHITE;        // Current color to move (turn).
   this.setup = C.SETUP_START;  // Current position. Array of char (p P k K .). Length 52.
   this.clicks = {};            // Store square clicks
   this.his = [];               // Store history: last item is previous position
   this.started = false;
   this.myColor = C.WHITE;  // Color of player where opponent is server.
   this.engine = "";        // Name of engine
   this.stopwatch = new StopWatch();  // To measure thinking time of a move
   this.stopwatch.start();
   this.setupmode = false;  // If true, manual setup by mouse clicks

   // === METHODS ===
   this.init = function(iSetup, iColor) {
      // Parameter iSetup has two appearances: list of char or string of char.
      // The string can differ in that it has also spaces to visualize the board.
      this.color = iColor;  // 0,1
      // If needed convert iSetup to array.
      if (iSetup.length == 52) {
         // No spaces in setup (string, maybe list)
         this.setup = Array.from(iSetup);  // make clone
      }
      else {
         // Remove spaces from setup (string, maybe list)
         var setup = Array.from(iSetup);  // make clone
         this.setup = setup.filter(function(item) { return item !== " " }); // Filter spaces.
      }
      this.clicks = {};
      this.his = [];
      this.stopwatch.start();
      this.xprint();
      updateFooter();
      return 0;
   }  // init()

   this.update = function(iSetup, iColor) {
      // Updates game setup and color. History unchanged.
      this.setup = iSetup;
      this.color = iColor;
      this.clicks = {};
      this.stopwatch.start();  // start count for next move
      this.xprint();
      updateFooter();
      return 0;
   }  // update()

   this.initFromFen = function(iFen) {
      // Set position with given fen string.
      var rlist = this.parseFEN(iFen);
      var sideToMove = rlist[0];

      // prepare output
      var pcode = {'w': 'P', 'W': 'K', 'b': 'p', 'B': 'k', '0': '.'};
      var board = ["0"];
      for (var i=1; i < rlist.length; i++) {
         board.push(pcode[ rlist[i] ]);
      }
      board.push("0");

      ///this.color = (sideToMove == "W") ? C.WHITE : C.BLACK; // alternatively
      var color = {W: C.WHITE, B: C.BLACK}[sideToMove];
      this.init(board, color);
      return 0;
   }  // initFromFen()

   this.newPosition = function() {
      // Get "one-color" position object of current position.
      if (this.color == C.WHITE) { 
         var pos = new Position(this.setup);
      }
      else {
         var rotSetup = this.setup.map(swapCase).reverse();  // clone
         var pos = new Position( rotSetup );
      }
      return pos;
   }  // newPosition()

   this.toFEN = function() {
      // Returns FEN string of current setup with colorToMove indication.
      var sideToMove = ['W', 'B'][this.color];
      var whitePieces = "";
      var first = true;
      for (var num=1; num < 51; num++) {
          var pieceCode = this.setup[num];
          if (pieceCode == 'P' && first == false)  whitePieces += ",";
          if (pieceCode == 'K' && first == false)  whitePieces += ",";
          if (pieceCode == 'P') { whitePieces += num.toString(); first = false; }
          if (pieceCode == 'K') { whitePieces += 'K' + num.toString(); first = false; }
      }
      var blackPieces = "";
      var first = true;
      for (var num=1; num < 51; num++) {
          var pieceCode = this.setup[num];
          if (pieceCode == 'p' && first == false)  blackPieces += ",";
          if (pieceCode == 'k' && first == false)  blackPieces += ",";
          if (pieceCode == 'p') { blackPieces += num.toString(); first = false; }
          if (pieceCode == 'k') { blackPieces += 'K' + num.toString(); first = false; }
      }
      var fenString = sideToMove + ":W" + whitePieces + ":B" + blackPieces + ".";
      return fenString;
   }  // toFEN()

   this.onSquareClick = function(mousePos, keysPressed) {
      // Called by mousedown on the canvas. Executes only if dark square selected.
      // Argument 'mousePos' is selected row/col
      if (game.setupmode == true)   this.tryToSetup(mousePos);
      else  this.tryToMove(mousePos);
      return 0;
   }  // onSquareClick()

   this.tryToMove = function(mousePos) {
      // Called by onSquareClick. Executes only if setupmode is false (movemode).
      // Argument 'mousePos' is selected row/col

      if (this.started == true && this.color != this.myColor) {
         alert("Wait for move from server.");
         return 0;
      }

      if (!this.clicks[mousePos.row])   this.clicks[mousePos.row] = {};  // init row
      if (this.clicks[mousePos.row][mousePos.col]) {
         // remove property if already present
         delete this.clicks[mousePos.row][mousePos.col];
      }
      else {
         // store property
         this.clicks[mousePos.row][mousePos.col] = true;
      }
      this.xprint();  // color clicked squares

      //this.testClicks();
      // Try to match a move with clicked squares.
      var numbers = this.clickedNumbers();
      var matchedMoves = this.matchMoves(numbers);

      //console.log("matchedMoves: ");
      //for (j in matchedMoves) {
      //    console.log("       " + matchedMoves[j].steps + " " + matchedMoves[j].takes);
      //}

      if (matchedMoves.length == 0) {
         var player = ["white","black"][this.color];
         alert("No move found: " + player + " must play.");
         this.clicks = {};
         return 0;
      }

      if (matchedMoves.length > 1) return 0;  // do nothing

      if (matchedMoves.length == 1) {
         // move if unique match found
         var move = matchedMoves[0];
         this.domove(move);
      }

      return 0;
   }  // tryToMove()

   this.tryToSetup = function(mousePos) {
      // Called by onSquareClick. Executes only if setupmode is true.
      // Argument 'mousePos' is selected row/col
      var number = rowcolToNumber(mousePos.row, mousePos.col);
      var pcode = {'w': 'P', 'W': 'K', 'b': 'p', 'B': 'k', '0': '.'};
      this.setup[number] = pcode[window.pieceSelector];
      this.xprint();
      return 0;
   }  // tryToSetup()

   this.clickedNumbers = function() {
      // Convert clicked row/col to square numbers.
      var numbers = [];      // clicked numbers
      for (var row in this.clicks) {
         for (var col in this.clicks[row]) {
            var nrow = parseInt(row);
            var ncol = parseInt(col);
            var number = rowcolToNumber(nrow, ncol);
            numbers.push(number);
         }
      }
      return numbers;
   }  // clickedNumbers()

   this.testClicks = function() {
      // retrieve clicks and report
      var str = "";
      for (var row in this.clicks) {
         for (var col in this.clicks[row]) {
            str += "[" + row + "," + col + "] ";
         }
      }
      console.log("CLICKS: " + str);
      return 0;
   }  // testClicks()

   this.legalMoves = function() {
      // Returns  "two-color" legal moves of position.
      var pos = this.newPosition();      // current "one-color" position
      var legalMoves = pos.legalMoves(); // "one-color" legal moves
      if (this.color == C.WHITE)  return legalMoves;

      // For C.BLACK: convert "one-color" moves to "two-color" moves
      var legalMoves2 = [];
      for (var i=0; i < legalMoves.length; i++) {
         var move = legalMoves[i];
         var steps2 = move.steps.map( function(el) { return (51-parseInt(el)) } );
         var takes2 = move.takes.map( function(el) { return (51-parseInt(el)) } );
         var move2 = new Move( {steps: steps2, takes: takes2} );
         legalMoves2.push(move2);
      }
      return legalMoves2;
   }  // legalMoves()

   this.matchMoves = function(iNumbers) {
      // iNumbers: "two-color" square numbers
      // Returns "two-color" moves which matches the given square numbers.
      var pos = this.newPosition();  // current "one-color" position

      var numbers = [];  // "one-color" square numbers
      for (var i=0; i < iNumbers.length; i++) {
         if (this.color == C.WHITE) numbers.push(iNumbers[i]);
         else numbers.push(51 - iNumbers[i]);
      }

      var matchedMoves = pos.matchMoves(numbers);  // "one-color" match moves
      if (this.color == C.WHITE)  return matchedMoves;

      // For C.BLACK: convert "one-color" moves to "two-color" moves
      var matchedMoves2 = [];
      for (var i=0; i < matchedMoves.length; i++) {
         var move = matchedMoves[i];
         var steps2 = move.steps.map( function(el) { return (51-parseInt(el)) } );
         var takes2 = move.takes.map( function(el) { return (51-parseInt(el)) } );
         var move2 = new Move( {steps: steps2, takes: takes2} );
         matchedMoves2.push(move2);
      }
      return matchedMoves2;
   }  // matchMoves()

   this.domove = function(iMove) {
      // "Two-color" counterpart of pos.domove; parameter iMove is a "two-color" move.
      // Called by user or used to move the engine.
      var pos = this.newPosition();  // current "one-color" position

      // Convert move to "one-color"  version
      if (this.color == C.WHITE) {
         var move = iMove;
      }
      else {
         var steps = iMove.steps.map( function(el) { return (51-parseInt(el)) } );
         var takes = iMove.takes.map( function(el) { return (51-parseInt(el)) } );
         var move = new Move( {steps: steps, takes: takes} );
      }

      var newpos = pos.domove(move);


      this.info("Last move: " + moveToString(iMove) );

      // add old position to history
      this.addHis(this.color, this.setup);

      this.stopwatch.stop();
      var timeSpend = Math.round(this.stopwatch.elapsedMilliSec/1000);  // seconds
      //console.log("Time spend sec: " + timeSpend);

      if (this.started == true && this.color == this.myColor) {
         sendDxpMove(iMove, timeSpend);
      }

      // Update new position
      if (this.color == C.WHITE)  var newSetup = newpos.setup;
      else  var newSetup = newpos.setup.map(swapCase).reverse();
      var newColor = 1 - this.color;   // switch to other color
      this.update(newSetup, newColor);
      return 0;
   }  // domove()

   this.backToMove = function(moveId, color) {
      // Set position from history back to given moveId (number) and color.
      var found = false;
      for (var idx=0; idx < this.his.length; idx++) {
         if (this.his[idx].moveId == moveId && this.his[idx].color == color) {
            found = true;
            break;
         }
      }
      //console.log("backToMove: ", moveId, color );
      if (!found) return 0;   // Cannot move back
      var newSetup = this.his[idx].setup;
      var newColor = this.his[idx].color;
      this.update(newSetup, newColor);     // Update new position
      this.his = this.his.slice(0, idx);   // Remove history from idx..end
      return 0;
   }  // backToMove()

   this.addHis = function(iColor, iSetup) {
      if (this.his.length == 0) {
         var nextMoveId = 1;
      }
      else {
         var last = this.his[this.his.length-1];
         if (last.color == C.WHITE) var nextMoveId = last.moveId;
         else  var nextMoveId = last.moveId + 1;
      }
      this.his.push({moveId: nextMoveId, color: iColor, setup: iSetup});
      return 0;
   }  // addHis()

   this.parseFEN = function(iFen) {
      // Parses a string in Forsyth-Edwards Notation into a Position
      var fen = iFen;                   // working copy
      fen = fen.replace(/\s+/, "");     // remove all spaces
      fen = fen.split(".")[0];          // cut off info (.xxx) at the end
      if (fen == "") fen = "W:B:W";     // empty FEN Position
      if (fen == "W::") fen = "W:B:W";
      if (fen == "B::") fen = "B:B:W";
      if (/^.::$/.test(fen)) fen = "W:B:W";
      var parts = fen.split(":");

      var rlist = Array(52).join("0").split("");  // init temp return list 0..50
      var sideToMove = (parts[0][0] == "B") ? "B" : "W";
      rlist[0] = sideToMove;

      for (i in { 1:0, 2:0 } ) { 
         // process the two sides
         var side = parts[i];
         var color = side[0];
         side = side.substring(1);  // strip color char
         if (side.length == 0) continue;

         var numSquares = side.split(",")   // list of numbers or range of numbers with/without king flag
         for (var i=0; i < numSquares.length; i++) {
            var num = numSquares[i];
            var isKing = (num[0] == "K") ? true : false;
            num = isKing ? num.substring(1) : num;    // strip 'K'
            var range = num.split("-");
            switch (range.length) {
               case 1:  // no range
                  rlist[parseInt(num)] = isKing ? color.toUpperCase() : color.toLowerCase();
                  break;
               case 2:  // range n-m
                  var from = parseInt(range[0]);
                  var to = parseInt(range[1]);
                  for (var j=from; j < (to + 1); j++ ) {
                     rlist[j] = isKing ? color.toUpperCase() : color.toLowerCase();
                  }
                  break;
               default:
                  continue;  // error; do nothing
            }
         }  // for num
      }  // for sides
      return rlist;
   }  // parseFEN()

   this.info = function(str) {
      var infoElement = document.getElementById("iPlayInfo");
      var escaped = str.
        replace(/&/, "&amp;").
        replace(/</, "&lt;").
        replace(/>/, "&gt;").
        replace(/"/, "&quot;");
      infoElement.value = escaped;
   }  // info()

   this.xprint = function() {
      //var canvas = document.getElementById('myCanvas'); 
      //var ctx = canvas.getContext('2d'); 
      var canvas = window.myCanvas.canvas;
      var ctx = window.myCanvas.ctx;

      var bw = C.BOARD_RECT.w;
      var bh = C.BOARD_RECT.h;
      var bx = C.BOARD_RECT.x;
      var by = C.BOARD_RECT.y;

      ctx.save();   // =====================================

      // Background color canvas
      ctx.fillStyle = C.BG_COLOR;
      ctx.fillRect(0, 0, canvas.width, canvas.height);

      // Draw light squares as one rect
      ctx.fillStyle = C.SQR_COLORLIGHT;
      ctx.fillRect(bx, by, bw, bh);

      // Draw dark squares
      for (var row = 0; row < 10; row++) {
         for (var col = 0; col < 10; col++) {
            if ((row + col) % 2 != 0) {
               var dx = bx + col * C.SQUARESIZE;
               var dy = by + row * C.SQUARESIZE;
               if (this.clicks[row]) {
                  ctx.fillStyle = this.clicks[row][col] ? "red" : C.SQR_COLORDARK; 
               }
               else {
                  ctx.fillStyle = C.SQR_COLORDARK; 
               }
               ctx.fillRect(dx, dy, C.SQUARESIZE, C.SQUARESIZE);
            }
         }
      }

      // Setup pieces
      for (var i = 1; i < this.setup.length; i++) {
         var piececode = this.setup[i];
         if (piececode != '0' && piececode != '.') {
            var piece = {number: i, code: piececode};
            this.drawPiece(ctx, piece);
         }
      }

      // Draw setup mode indication
      if (this.setupmode == true) {
         ctx.fillStyle = "red";
         ctx.globalAlpha = 0.2;   // 0.0 .. 1.0    (from transparent to opaque)
         ctx.fillRect(0, 0, canvas.width, canvas.height);
         ctx.globalAlpha = 1.0;
      }

      ctx.restore();   // =====================================
      return 0;
   }  // xprint

   this.drawPiece = function(ctx, piece) {
      var origin = {x: C.BOARD_RECT.x, y: C.BOARD_RECT.y};
      var loc = numberToRowCol(piece.number);       
      var dx = origin.x + (loc.col * C.SQUARESIZE);
      var dy = origin.y + (loc.row * C.SQUARESIZE);
      var dw = C.SQUARESIZE; 
      var dh = C.SQUARESIZE;
      var sx = C.VIEWS[piece.code].sx;
      var sy = C.VIEWS[piece.code].sy;
      var sw = C.VIEWS[piece.code].sw;
      var sh = C.VIEWS[piece.code].sh;
      ctx.drawImage(C.PIECE_IMAGE, sx, sy, sw, sh, dx, dy, dw, dh);
      return 0;
   }  // drawPiece

   return this; // =================
}  // Game()

// ====================================================================================



include/js/wsdam_position.js
/*
||====================================================================================  
||  WsDam: browser client for 10x10 draughts
||====================================================================================
||  Module position related stuff.
||  ALL LOGIC FOR POSITION IS "ONE-COLOR" LOGIC: white plays.
||  Numbering of board from top-left to bottom-right.
||
|| (c) Copyright 2018 Arthur Kalverboer 
||
||====================================================================================
*/

Position = function(iSetup) {
   // A position of a draughts 10x10 game
   // Position stored as an array of 52 char; first and last index unused ('0') rotation-symmetry
   // Coding:
   // - lowercase char: black
   // - uppercase char: white
   // - P: white piece; K: white king
   // - p: black piece; k: black king
   // - dot is empty field
   // The setup is a "one-color" setup. Always white to move.
   // =============================================================================

   this.setup = Array.from(iSetup);  // make clone

   // === METHODS ===
   this.rotate = function() {
      // Swap case of each char in the setup (black <--> white) and reverse setup order.
      var rotSetup = this.setup.map(swapCase).reverse();  // clone
      var rotpos = new Position( rotSetup );
      return rotpos;
   }  // rotate()

   this.clone = function() {
      var setup = Array.from(this.setup);  // make clone
      var newpos = new Position(setup);
      return newpos
   }  // clone()

   this.legalMoves = function() {
      return mv.gen_moves(this.setup);
   }  // legalMoves()

   this.matchMoves = function(iNumbers) {
      // Returns array of moves which matches with given square numbers.
      // Split numbers in steps and takes.
      var steps = [];  var takes = [];
      for (var i=0; i < iNumbers.length; i++) {
         if ( isLowerCase(this.setup[iNumbers[i]]) ) takes.push(iNumbers[i]);
         else steps.push(iNumbers[i]);
      }

      var matchedMoves = [];
      var legalMoves = this.legalMoves();
      for (var i=0; i < legalMoves.length; i++) {
         lmove = legalMoves[i];
         //var matchFound = mv.matchMove(lmove, nsteps, ntakes);
         var matchFound = lmove.match(steps, takes);
         // *** Special match if first and last step equal.***
         var first = lmove.steps[0]; var last = lmove.steps[lmove.steps.length-1];
         var isSpecialMatch = ( steps.length == 1 && first == steps[0] && last == steps[0] );
         if (isSpecialMatch) return [lmove];

         if (matchFound) matchedMoves.push(lmove);
      }
      return matchedMoves;
   }  // matchMoves()

   this.hasCapture = function() {
      return mv.hasCapture(this.setup);
   }  // hasCapture()

   this.domove = function(move) {
      // Move is an object with steps and takes.
      // The move is a "one-color" move; white always playing.
      // Returns new position object after moving.
      // This function does not verify if the move is legal.
      if (move == null)  return this; 
      var setup = this.setup.slice(0);    // clone setup

      // Actual move
      var i = move.steps[0];      // first
      var j = move.steps[move.steps.length-1]; // last (NB. sometimes i==j !)
      var p =  setup[i];

      // Move piece and promote to white king
      setup[i] = '.';
      var promotion_line = [1,2,3,4,5];
      if (promotion_line.indexOf(j) > -1 && p!= 'K')  setup[j] = 'K';
      else setup[j] = p;

      // Capture
      for (var k=0; k < move.takes.length; k++) setup[move.takes[k]] = '.';

      var posnew = new Position(setup);
      return posnew;
   }  // doMove()

   return this;
}  //===  Position


// ====================================================================================



include/js/wsdam_config.js
/*
||====================================================================================  
||  WsDam: browser client for 10x10 draughts
||====================================================================================
||  CONSTANTS and CONFIG
||
|| (c) Copyright 2018 Arthur Kalverboer 
||
||====================================================================================
*/

C = {}  // global constants
C.VERSION = "2018.04.30";
C.INITIATOR = "WsDam " + "(" + C.VERSION + ")";   // name DXP protocol
C.APPNAME = {short:"WsDam", long:"WsDam Draughts Client", github: "wsdam_draughts_client"};
C.DESCR = "Webclient for 10x10 draughts";

C.WHITE = 0;
C.BLACK = 1;
C.BOARD_DIM = 10;  // fixed 10x10 board (do not change!)
C.SQUARESIZE = 42; // can be changed for larger board
C.SQR_COLORLIGHT = "#BFBAB6";  
C.SQR_COLORDARK = "#7E858C"
C.BG_COLOR = "#50545D";     // Background color canvas

var w = h = C.SQUARESIZE * C.BOARD_DIM;
C.BOARD_RECT = {x: 10, y: 10, w: w, h: h};  // to fill the canvas
C.CANVAS_SIZE = {w: C.BOARD_RECT.w + 20, h: C.BOARD_RECT.h + 20};
C.PIECE_IMAGE = document.createElement('IMG');

// Choose one of the sprites to display pieces
  //C.PIECE_IMAGE.src = "include/images/spr_th01.png"
  C.PIECE_IMAGE.src = "include/images/spr_th60.svg"
  //C.PIECE_IMAGE.src = "include/images/spr_th61.svg"
  //C.PIECE_IMAGE.src = "include/images/spr_th01.svg"

// Views for sprite images of pieces: sequence bk, bm, wk, wm
C.VIEWS = { k:{}, p:{}, K:{}, P:{} }
C.VIEWS.k = {sx: 0, sy: 0, sw: 48, sh: 48};    // bk
C.VIEWS.p = {sx: 48, sy: 0, sw: 48, sh: 48};   // bm
C.VIEWS.K = {sx: 96, sy: 0, sw: 48, sh: 48};   // wk
C.VIEWS.P = {sx: 144, sy: 0, sw: 48, sh: 48};  // wm

C.SETUP_START = [];  // array of starting position
C.SETUP_START.push('0');
for (var i=1; i<=20; i++)  C.SETUP_START.push('p');
for (var i=21; i<=30; i++) C.SETUP_START.push('.');
for (var i=31; i<=50; i++) C.SETUP_START.push('P');
C.SETUP_START.push('0');

// The external respresentation of our board is a character string.
// Sequence of chars (P K p k .) with 0 at both ends and spaces for visualisation.
BOARD_EMPTY = 
"0 \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
0" // ==========================

BOARD_START =
"0 \
        p   p   p   p   p      \
      p   p   p   p   p        \
        p   p   p   p   p      \
      p   p   p   p   p        \
        .   .   .   .   .      \
      .   .   .   .   .        \
        P   P   P   P   P      \
      P   P   P   P   P        \
        P   P   P   P   P      \
      P   P   P   P   P        \
0" // ==========================



BOARD_TEST_01 =
"0 \
        .   K   .   .   .      \
      .   .   .   .   .        \
        p   .   .   .   .      \
      .   .   .   k   p        \
        p   .   .   .   p      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   p   .   .        \
        .   .   .   p   .      \
      .   .   .   .   .        \
0" // ==========================

BOARD_TEST_02 =
"0 \
        .   K   .   .   .      \
      .   .   .   .   .        \
        p   .   .   .   .      \
      .   .   .   k   p        \
        p   .   .   .   .      \
      .   .   .   .   p        \
        .   .   .   .   .      \
      .   .   p   .   .        \
        .   .   .   p   .      \
      .   .   .   .   .        \
0" // ==========================


BOARD_TEST_03 =
"0 \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
        p   p   .   .   .      \
      P   .   .   .   .        \
        p   p   p   p   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
0" // ==========================

BOARD_TEST_04 =
"0 \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   p   .   p   .        \
        .   .   .   .   .      \
      K   .   .   .   .        \
        .   .   .   .   .      \
      .   p   .   p   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
0" // ==========================

BOARD_TEST_05 =
"0 \
        .   .   .   .   .      \
      .   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   p   .        \
        .   p   p   .   .      \
      .   .   .   .   .        \
        .   .   p   p   .      \
      K   .   .   .   .        \
        .   .   .   .   .      \
      .   .   .   .   .        \
0" // ==========================


BOARD_PROBLEM_01 =  // P.Lauwen, DP, 4/1977
"0 \
        .   .   .   .   p      \
      .   .   p   .   .        \
        .   .   .   .   P      \
      .   .   .   P   .        \
        .   .   .   P   .      \
      .   .   .   P   p        \
        .   P   .   .   p      \
      .   p   .   .   p        \
        P   p   .   .   p      \
      .   .   .   P   P        \
0" // ==========================

// FEN examples
FEN_INITIAL = "W:B1-20:W31-50";
FEN_DXP100_1 = "W:W15,19,24,29,32,41,49,50:B5,8,30,35,37,40,42,45.";  // P.Lauwen, DP, 4/1977
FEN_DXP100_2 = "W:W17,28,32,33,38,41,43:B10,18-20,23,24,37.";
FEN_DXP100_3 = "W:WK3,25,34,45:B38,K47.";
FEN_DXP100_4 = "W:W18,23,31,33,34,39,47:B8,11,20,24,25,26,32.";       // M.Dalman
FEN_DXP100_5 = "B:B7,11,13,17,20,22,24,30,41:W26,28,29,31,32,33,38,40,48.";  // after 30-35 white wins
FEN_DXP100_6 = "W:W30,32,34,35,37,39,40,42,45,46,47,49:B1,8,10,11,12,14,17,19,20,21,23,26.";  // Lucien Lebourg

// Solution 1x1 after 20 moves!! Mad100 finds solution. Set nodes 300000.
FEN_DXP100_7 = "W:W16,21,25,32,37,38,41,42,45,46,49,50:B8,9,12,17,18,19,26,29,30,33,34,35,36.";



// ====================================================================================

